src/table.js

```diff
@@ -666,3 +667,90 @@ export function __table(source, operations) {
   }
   return source;
 }
+
+function isValidSchema(schema) {
+  if (!schema || !Array.isArray(schema)) return;
+  return schema.every((s) => {
+    s && typeof s.name === "string" && typeof s.type === "string";
+  });
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0, // TODO for csv, tsv?
+    buffer: 0
+  };
+}
+
+function inferSchema(source) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  if (arrayIsPrimitive(sample)) {
+    sample = sample.map((d) => {
+      return {value: d};
+    });
+  }
+  const typeCounts = {};
+  sample.map((d) => {
+    for (const key in d) {
+      if (!typeCounts[key]) typeCounts[key] = initKey();
+      // for json and sqlite, we already have some types, but for csv and tsv, all
+      // columns are strings here.
+      const type = typeof d[key];
+      const value = type === "string" ? d[key]?.trim() : d[key];
+      if (value === null || value === undefined || value.length === 0)
+        typeCounts[key]["other"]++;
```

Comment:
Not sure it’s appropriate to consider null/undefined/empty to be “other”. I would consider a column with 80% null/undefined and 20% string to be type string. In other words instead of counting these empty/missing values as a type, we should ignore them and only count types of present values. Otherwise sparsely populated columns will be considered “other” which is likely undesirable.

(Perhaps this is an alternative to the special treatment of “other” below, but I think for values that are truly “other” then we wouldn’t want that special treatment; the special treatment is only needed because we are considering nullish/empty values as “other” here.)